### PR TITLE
Change cursor on context clicks on XY-charts

### LINKED ISSUE #636: Change cursor when doing right-mouse drag zoom in XY-charts
In comparison to the Time Graph charts, the cursor is not changed  in xy-charts when the user is doing zoom operation using `right-mouse click drag`.

missing-right-mouse-drag-cursor

### PR DESCRIPTION
**Pull request by Rodrigoplp-work**
Change the cursor when dragging with the left button (crosshair cursor)
and with the right button (ew-resize cursor).

The behavior of the xy-charts now is the same as on time-graph charts.

Fixes #636.

Signed-off-by: Rodrigo Pinto <rodrigo.pinto@calian.ca>

### CODE REVIEW

**Review by PatrickTasse — CHANGES_REQUESTED** [review]
Not sure if these comments are in the scope of this PR or if they need a new PR:

There are some small differences in behavior compared with time-graph charts:

- If you start dragging in one mode (zoom or select) then click the other button, it starts a new action with the new mode. The other buttons should be ignored while an action is already in progress.
- In time-graph chart, holding the Shift button will change to the crosshair cursor to show that a click will extend the selection instead of starting a new selection. Note that pressing any other modifier key will cancel this mode.

**Comment by Rodrigoplp-work** [comment]
- Shift button: implemented;
- Avoid click while dragging: I made a few attempts but didn't succeed. The problem is not only on the xy-chart, it also happens on others, like in the Resources Status chart. If we really want to do this, I suggest we open a new ticket to address it in the future.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]
If you hold Shift, then click to extend selection, if the Shift is still being pressed then the cursor should still be the crosshair.

About the click while dragging, I'm not sure if it's the same use case that was tried because it doesn't happen with time graphs:
- Left mouse down (starts time selection)
- Mouse move (extends time selection)
- Right click (ignored, nothing happens) <- but xy-chart different
- Mouse move (continues to extend time selection)
- Left mouse up (finishes time selection)
And same thing when swapping left and right buttons.
But it's OK to address this separately, as it's not only related to the cursor but to the action handling itself also.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 1]
```diff
@@ -797,10 +803,20 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                    this.setState({ cursor: 'crosshair' });
+                    break;
+                }
             }
         }
     }
 
+    private onKeyUp(key: React.KeyboardEvent) {
+        if (key.key === 'Shift') {
+            this.setState({ cursor: 'default' });
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:816]: It should not reset the cursor is selection is in progress with the left mouse button being pressed.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 2]
```diff
@@ -797,10 +803,20 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                     this.pan(PAN_RIGHT);
                     break;
                 }
+                case 'Shift': {
+                    this.setState({ cursor: 'crosshair' });
```
Comment (PatrickTasse) [inline packages/react-components/src/components/xy-output-component.tsx:807]: It should not change the cursor if zoom selection is in progress with the right mouse button being pressed.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 1 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx]: Fixed.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 2 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx]: Fixed.

**Review by PatrickTasse — DISMISSED** [review]

File: packages/react-components/src/components/xy-output-component.tsx [hunk 3]
```diff
@@ -71,6 +73,9 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
            this.applySelectionZoom();
         }
         this.mouseIsDown = false;
+        if (!this.state.shiftKey) {
+            this.setState({ cursor: 'default '});
```
Comment (marco-miller) [inline packages/react-components/src/components/xy-output-component.tsx:77]: Extra space right after "default".

File: packages/react-components/src/components/xy-output-component.tsx [hunk 4]
```diff
@@ -90,7 +95,9 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
             xyData: {},
             columns: [{title: 'Name', sortable: true}],
             allMax: 0,
-            allMin: 0
+            allMin: 0,
+            cursor: 'default',
```
Comment (marco-miller) [inline packages/react-components/src/components/xy-output-component.tsx:99]: Should a constant be defined and reused for this repeating 'default' value?

File: packages/react-components/src/components/xy-output-component.tsx [hunk 5]
```diff
@@ -797,7 +807,23 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
                     this.pan(PAN_RIGHT);
                     break;
                 }
+                case 'Shift': {
+                    if (!this.isRightClick) {
+                        this.setState({ cursor: 'crosshair', shiftKey: true });
```
Comment (marco-miller) [inline packages/react-components/src/components/xy-output-component.tsx:812]: Same question for this other repeated value.

File: packages/react-components/src/components/xy-output-component.tsx [hunk 6]
```diff
@@ -797,7 +807,23 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
...
+                    }
+                    break;
+                }
+            }
+        }
+    }
+
+    private onKeyUp(key: React.KeyboardEvent) {
+        if (key.key === 'Shift') {
```
Comment (marco-miller) [inline packages/react-components/src/components/xy-output-component.tsx:821]: Same question again, here for this key string value. Not sure if such duplication is done elsewhere or treated as normal (or expected).

**Review by marco-miller — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/xy-output-component.tsx (hunk 4 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx:99]: Fixed.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 3 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx]: Fixed.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 6 above)
Comment (Rodrigoplp-work) [reply packages/react-components/src/components/xy-output-component.tsx:821]: I would rather leave as is, as explained above.

File: packages/react-components/src/components/xy-output-component.tsx (hunk 4 above)
Comment (marco-miller) [reply packages/react-components/src/components/xy-output-component.tsx:99]: How about CURSOR_DEFAULT and so on, or one constant per reused cursor string value?

**Review by marco-miller — CHANGES_REQUESTED** [review]
Another proposal in case it could fit in; thanks.

**Comment by PatrickTasse** [comment]
I can understand Marco's point coming from Java where pretty much every code literal should be defined as a constant, as a good coding practice, especially when it is a fixed string defined in an API. If this was a non-sucking language it would provide API-defined constants for these CSS strings so that the client could use them directly...

But it's not just cursor constants in the same situation, there's also mouse event types 'mouseup', keyboard event key 'Shift', etc. We seem to be using literals for these all over the project (and other HTML projects?), so I'm not sure if we should start doing it differently here?

I'm fine either way.

**Review by PatrickTasse — APPROVED** [review]

**Review by marco-miller — APPROVED** [review]

**Pull request merged**